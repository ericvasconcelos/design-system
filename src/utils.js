export { arrayCombinations } from 'utils/arrayCombinations'
export { getBackgroundColor } from 'utils/getBackgroundColor'
export { getBorder } from 'utils/getBorder'
export { getButtonStyle } from 'utils/getButtonStyle'
export { getBorderRadius } from 'utils/getBorderRadius'
export { getColorFromName } from 'utils/getColorFromName'
export { getGuttersNames } from 'utils/getGuttersNames'
export { getIconNames } from 'utils/getIconNames'
export { getPadding } from 'utils/getPadding'
export { getPaddingNames } from 'utils/getPaddingNames'
export { getTypography } from 'utils/getTypography'
export { parsePalleteName } from 'utils/parsePalleteName'
export { styleColorValidator } from 'utils/styleColorValidator'
export { useAriaProps } from 'utils/useAriaProps'

export { DateUtils } from 'utils/DateUtils'
export { QueryUtils } from 'utils/QueryUtils'
export { ObjectUtils }  from 'utils/ObjectUtils'
export { CookiesUtils } from 'utils/CookiesUtils'
export { MoneyUtils } from 'utils/MoneyUtils'
export { NumberUtils } from 'utils/NumberUtils'