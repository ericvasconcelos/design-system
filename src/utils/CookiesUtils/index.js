export class CookiesUtils {
  static toObject(string = '') {
    const cookies = {}
    if (string !== '') {
      const parts = string.split('; ')
      parts.forEach(part => {
        const [key, value] = part.split('=')
        cookies[key] = value
      })
    }
    return cookies
  }
}
