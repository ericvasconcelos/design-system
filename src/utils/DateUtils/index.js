export class DateUtils {
  static toDate(string, format = 'pt-br') {
    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = string?.split('/')
        break
      case 'iso':
        [year, month, day] = string?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    return new Date(year, parseInt(month, 10) - 1, day)
  }

  static getAge(date) {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const difference = Date.now() - date.getTime()

    return Math.abs(new Date(difference).getUTCFullYear() - 1970)
  }

  static toString(date, format = 'pt-br') {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()
    const month = date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1
    const year = date.getFullYear()

    switch (format) {
      case 'pt-br':
        return `${day}/${month}/${year}`
      case 'iso':
        return `${year}-${month}-${day}`
      default:
        throw new Error('Format invalid')
    }
  }

  static validate(date, format = 'pt-br') {
    const monthDaysMap = {
      '01': 31,
      '02': 29,
      '03': 31,
      '04': 30,
      '05': 31,
      '06': 30,
      '07': 31,
      '08': 31,
      '09': 30,
      10: 31,
      11: 30,
      12: 31
    }
    let isValid = false

    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = date?.split('/')
        break
      case 'iso':
        [year, month, day] = date?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    isValid = isValid || parseInt(day, 10) <= monthDaysMap[month]
    isValid = isValid && parseInt(year, 10) > 1900

    return isValid
  }

  static convertMonthToYear(month) {
    return parseInt(month, 10) / 12
  }

  static formatPeriod(period, periodType) {
    const periodTypeMap = {
      monthly: { singular: 'mÃªs', plural: 'meses' },
      yearly: { singular: 'ano', plural: 'anos' },
      daily: { singular: 'dia', plural: 'dias' }
    }
    return `${period} ${periodTypeMap[periodType][period < 2 ? 'singular' : 'plural']}`
  }
}
