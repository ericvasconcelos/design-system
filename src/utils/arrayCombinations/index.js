export const arrayCombinations = (...args) => {
  const returnable = []
  const max = args.length - 1
  function helper(arr, i) {
    for (let j = 0, { length } = args[i]; j < length; j++) {
      const newArray = arr.slice(0) // clone arr
      newArray.push(args[i][j])
      if (i === max) returnable.push(newArray)
      else helper(newArray, i + 1)
    }
  }
  helper([], 0)
  return returnable
}
