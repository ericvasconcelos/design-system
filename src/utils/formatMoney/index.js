import { toMoney, toNumber } from 'vanilla-masker'

export const formatMoney = entry => {
  const moneyFormatter = {
    precision: 2,
    separator: ',',
    delimiter: '.',
    unit: 'R$'
  }

  return toMoney(toNumber(entry), moneyFormatter)
}
