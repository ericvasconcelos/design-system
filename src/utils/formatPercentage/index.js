import { toMoney, toNumber } from 'vanilla-masker'

export const formatPercentage = entry => {
  const percentageFormatter = {
    precision: 1,
    separator: ',',
    suffixUnit: '%'
  }
  return toMoney(toNumber(entry), percentageFormatter)
}
