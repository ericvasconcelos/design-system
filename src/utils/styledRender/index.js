import React from 'react'

import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { theme } from 'styles/theme'

export const styledRender = (children, options = {}) => {
  const rendered = render(<ThemeProvider theme={theme}>{children}</ThemeProvider>, options)
  return {
    ...rendered,
    rerender: (ui, rerenderOptions = {}) =>
      styledRender(ui, { container: rendered.container, ...rerenderOptions })
  }
}
