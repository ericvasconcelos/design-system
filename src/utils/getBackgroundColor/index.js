import { getColorFromName } from 'utils/getColorFromName'

export const getBackgroundColor = ({ theme, styleColor }) =>
  `background-color: ${getColorFromName({ theme, styleColor })};`
