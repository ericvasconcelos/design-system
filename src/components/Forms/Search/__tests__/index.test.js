import React, { useRef } from 'react'

import userEvent from '@testing-library/user-event'
import { styledRender } from 'utils/styledRender'

import { Search } from '../index.js'

const options = [
  [
    'value1',
    'key1',
    []
  ],
  [
    'value2',
    'key2',
    []
  ],
  [
    'value3',
    'key3',
    []
  ],
  [
    'value4',
    'key4',
    []
  ]
]


const Component = () => {
  const ref = useRef()

  return (
    <Search
      name="search"
      label="search"
      id="id"
      ariaLabel="search component"
      icon="search"
      options={options}
      ref={ref}
    />
  )
}

describe('Should render Search component when', () => {
  it('has default props', () => {
    const { getByTestId, getByText, getByLabelText } = styledRender(<Component />)

    userEvent.type(getByLabelText('search'), '1')
    userEvent.click(getByText('value1'))

    expect(getByTestId('Search')).toMatchSnapshot()
  })

  it('do not have any valid option', () => {
    const { getByTestId, getByText, getByLabelText } = styledRender(<Component />)

    userEvent.type(getByLabelText('search'), '9')

    expect(getByText('Não há nenhum resultado para a busca')).toBeInTheDocument()
    expect(getByTestId('Search')).toMatchSnapshot()
  })
})

