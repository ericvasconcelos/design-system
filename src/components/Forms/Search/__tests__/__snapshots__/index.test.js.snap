// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Search component when do not have any valid option 1`] = `
.c4 {
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/search.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/search.svg);
  background-color: #003FF2;
  background-image: none;
  background-size: 24px;
  width: 24px;
  height: 24px;
  -webkit-mask-position-x: 0;
  mask-position-x: 0;
  -webkit-mask-size: 24px;
  mask-size: 24px;
  -webkit-mask-repeat: no-repeat;
  mask-repeat: no-repeat;
}

.c6 {
  opacity: 1;
}

.c3 {
  border-radius: 8px;
  border: 1px solid rgba(51,51,51,0.1);
  border-color: #FFFFFF;
  background: #FFFFFF;
  color: #69788C;
  height: 56px;
  padding: 12px 0 0 15px;
  width: 100%;
  box-sizing: border-box;
}

.c3 ~ .c5 {
  position: absolute;
  right: 16px;
  top: 16px;
}

.c3:focus {
  outline: none !important;
  border: 1px solid #8DC6FF;
  box-shadow: 0 0 0 30px #FFFFFF inset, 0 0 0 4px #FFFFFF;
  box-shadow: 0px 0px 0px 3px #8DC6FF29;
}

.c3::-webkit-input-placeholder {
  color: transparent;
}

.c3::-moz-placeholder {
  color: transparent;
}

.c3:-ms-input-placeholder {
  color: transparent;
}

.c3::placeholder {
  color: transparent;
}

.c3 [type="search"]::-webkit-search-decoration,
.c3 [type="search"]::-webkit-search-cancel-button,
.c3 [type="search"]::-webkit-search-results-button,
.c3 [type="search"]::-webkit-search-results-decoration {
  -webkit-appearance: none;
}

.c2 {
  font-size: 0.5rem;
  position: absolute;
  left: 16px;
  top: 8px;
  pointer-events: none;
  color: #A6B5BB;
  -webkit-transition: 0.3s ease all;
  transition: 0.3s ease all;
  font-weight: bolder;
  text-transform: uppercase;
  opacity: 1;
}

.c0 {
  position: relative;
  max-width: 720px;
  margin: 16px 0;
  line-height: 16px;
}

.c0:focus-within .c1 {
  top: 8px;
  font-weight: bolder;
  color: #003FF2;
  font-size: 8px;
  text-transform: uppercase;
}

<div
  class="c0"
  data-testid="Search"
>
  <label
    class="c1 c2"
    for="id"
  >
    search
  </label>
  <input
    aria-label="search component"
    autocomplete="off"
    class="c3"
    id="id"
    name="search"
    value="9"
  />
  <div
    aria-hidden="true"
    aria-label=""
    class="c4 c5 c6"
    data-testid="Icon"
  />
</div>
`;

exports[`Should render Search component when has default props 1`] = `
.c4 {
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/search.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/search.svg);
  background-color: #A6B5BB;
  background-image: none;
  background-size: 24px;
  width: 24px;
  height: 24px;
  -webkit-mask-position-x: 0;
  mask-position-x: 0;
  -webkit-mask-size: 24px;
  mask-size: 24px;
  -webkit-mask-repeat: no-repeat;
  mask-repeat: no-repeat;
}

.c6 {
  opacity: 1;
}

.c3 {
  border-radius: 8px;
  border: 1px solid rgba(51,51,51,0.1);
  border-color: #FFFFFF;
  background: #FFFFFF;
  color: #69788C;
  height: 56px;
  padding: 12px 0 0 15px;
  width: 100%;
  box-sizing: border-box;
}

.c3 ~ .c5 {
  position: absolute;
  right: 16px;
  top: 16px;
}

.c3:focus {
  outline: none !important;
  border: 1px solid #8DC6FF;
  box-shadow: 0 0 0 30px #FFFFFF inset, 0 0 0 4px #FFFFFF;
  box-shadow: 0px 0px 0px 3px #8DC6FF29;
}

.c3::-webkit-input-placeholder {
  color: transparent;
}

.c3::-moz-placeholder {
  color: transparent;
}

.c3:-ms-input-placeholder {
  color: transparent;
}

.c3::placeholder {
  color: transparent;
}

.c3 [type="search"]::-webkit-search-decoration,
.c3 [type="search"]::-webkit-search-cancel-button,
.c3 [type="search"]::-webkit-search-results-button,
.c3 [type="search"]::-webkit-search-results-decoration {
  -webkit-appearance: none;
}

.c2 {
  font-size: 0.5rem;
  position: absolute;
  left: 16px;
  top: 8px;
  pointer-events: none;
  color: #A6B5BB;
  -webkit-transition: 0.3s ease all;
  transition: 0.3s ease all;
  font-weight: bolder;
  text-transform: uppercase;
  opacity: 1;
}

.c0 {
  position: relative;
  max-width: 720px;
  margin: 16px 0;
  line-height: 16px;
}

.c0:focus-within .c1 {
  top: 8px;
  font-weight: bolder;
  color: #003FF2;
  font-size: 8px;
  text-transform: uppercase;
}

<div
  class="c0"
  data-testid="Search"
>
  <label
    class="c1 c2"
    for="id"
  >
    search
  </label>
  <input
    aria-label="search component"
    autocomplete="off"
    class="c3"
    id="id"
    name="search"
    value="value1"
  />
  <div
    aria-hidden="true"
    aria-label=""
    class="c4 c5 c6"
    data-testid="Icon"
  />
</div>
`;
